File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/State/TurnStateEntry.cs
using Microsoft.Teams.AI.Utilities;
using System.Runtime.CompilerServices;
using System.Text.Json;

namespace Microsoft.Teams.AI.State
{
    /// <summary>
    /// Accessor class for managing an individual state scope.
    /// </summary>
    public class TurnStateEntry
    {
        private Record _value;
        private string _hash;
        private static readonly JsonSerializerOptions _serializerOptions = new() { MaxDepth = 64 };

        /// <summary>
        /// Constructs the turn state entry.
        /// </summary>
        /// <param name="value">Value to initialize the state scope with. The default is an {} object.</param>
        /// <param name="storageKey">Storage key to use when persisting the state scope.</param>
        public TurnStateEntry(Record value, string? storageKey = null)
        {
            Verify.ParamNotNull(value);
            _value = value;
            StorageKey = storageKey;
            _hash = ComputeHash(value);
        }

        /// <inheritdoc />
        public bool HasChanged
        {
            get { return ComputeHash(_value!) != _hash; }
        }

        /// <inheritdoc />
        public bool IsDeleted { get; private set; } = false;

        /// <inheritdoc />
        public Record? Value
        {
            get
            {
                if (IsDeleted)
                {
                    _value = new();
                    IsDeleted = false;
                }

                return _value;
            }
        }

        /// <inheritdoc />
        public string? StorageKey { get; }

        /// <summary>
        /// Clears the state scope.
        /// </summary>
        public void Delete()
        {
            IsDeleted = true;
        }

        /// <summary>
        /// Replaces the state scope with a new value.
        /// </summary>
        /// <param name="value">New value to replace the state scope with.</param>
        public void Replace(Record value)
        {
            Verify.ParamNotNull(value);
            _value = value;
        }

        // TODO: Optimize if possible
        /// <summary>
        /// Computes the hash from the object
        /// </summary>
        /// <param name="obj">The object to compute has from</param>
        /// <returns>Returns a Json object representation </returns>
        internal static string ComputeHash(object obj)
        {
            Verify.ParamNotNull(obj);

            return JsonSerializer.Serialize(obj, _serializerOptions);
        }
    }
}

File: python/packages/ai/teams/state/turn_state_entry.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import Any, Dict, Iterator, MutableMapping, Optional


class TurnStateEntry(MutableMapping[str, Any]):
    """Accessor class for managing an individual state scope."""

    _storage_key: Optional[str]
    _deleted: bool = False
    _hash: str
    data: Dict[str, Any]

    def __init__(self, value: Optional[Dict[str, Any]] = None, storage_key: Optional[str] = None):
        """Creates a new instance of the `TurnStateEntry` class.

        Args:
            value (Optional[Dict[str, Any]], optional): Value to initialize the state scope with.
                Defaults to None which means creating an empty dict.
            storage_key (Optional[str], optional): Storage key to use when persisting the
                state scope. Defaults to None.
        """
        super().__init__()
        self.data = value or {}
        self._storage_key = storage_key
        self._hash = str(self.data)

    @property
    def has_changed(self) -> bool:
        """Gets a value indicating whether the state scope has changed since it was last loaded.

        Returns:
            bool: True if the value has changed, False otherwise.
        """
        return str(self.data) != self._hash

    @property
    def is_deleted(self) -> bool:
        """Gets a value indicating whether the state scope has been deleted.

        Returns:
            bool: True if the state entry is deleted, False otherwise.
        """
        return self._deleted

    @property
    def storage_key(self) -> Optional[str]:
        """Gets the storage key used to persist the state scope.

        Returns:
            Optional[str]: The storage key used to persist the state scope.
        """
        return self._storage_key

    def delete(self) -> None:
        """Clears the state scope."""
        self.data = {}
        self._deleted = True

    def replace(self, value: Optional[Dict[str, Any]] = None) -> None:
        """Replaces the state scope with a new value.

        Args:
            value (Optional[Dict[str, Any]], optional): New value to replace the state scope with.
        """
        self.data = {}
        self._deleted = False

        if not value:
            return None

        self.data = value
        return None

    def __getitem__(self, key: str) -> Optional[Any]:
        self._deleted = False
        return self.data.get(key)

    def __setitem__(self, key: str, value: Any) -> None:
        self._deleted = False
        self.data[key] = value

    def __contains__(self, key: object) -> bool:
        return key in self.data

    def __delitem__(self, key: str) -> None:
        if key in self.data:
            del self.data[key]

    def __iter__(self) -> Iterator[str]:
        return iter(self.data)

    def __len__(self) -> int:
        return len(self.data)

    def __getattribute__(self, name: str) -> Any:
        if name == "data":
            self._deleted = False

        return super().__getattribute__(name)
